annotated.go

```diff
@@ -272,210 +529,357 @@ func (la *lifecycleHookAnnotation) apply(ann *annotated) error {
+				field := t.Field(j)
+				if field.Type == _typeOfContext && ctxPos < 0 {
+					ctxStructPos = i
+					ctxPos = j
+					continue
+				}
+				fields = append(fields, field)
+			}
+			invokeParamTypes = append(invokeParamTypes, reflect.StructOf(fields))
+			continue
 		}
+		invokeParamTypes = append(invokeParamTypes, origHookFnT.In(i))
 	}
+	invokeFnT := reflect.FuncOf(invokeParamTypes, []reflect.Type{}, false)
+	invokeFn := reflect.MakeFunc(invokeFnT, func(args []reflect.Value) (results []reflect.Value) {
+		lc := args[0].Interface().(Lifecycle)
+		args = args[1:]
+		hookArgs := make([]reflect.Value, origHookFnT.NumIn())
 
-	return
-}
+		hookFn := func(ctx context.Context) (err error) {
+			// If the hook function has multiple parameters, and the first
+			// parameter is a context, inject the provided context.
+			if ctxStructPos < 0 {
+				offset := 0
+				for i := 0; i < len(hookArgs); i++ {
+					if i == ctxPos {
+						hookArgs[i] = reflect.ValueOf(ctx)
+						offset = 1
+						continue
+					}
+					if i-offset >= 0 && i-offset < len(args) {
+						hookArgs[i] = args[i-offset]
+					}
+				}
+			} else {
+				for i := 0; i < origHookFnT.NumIn(); i++ {
+					if i != ctxStructPos {
+						hookArgs[i] = args[i]
+						continue
+					}
+					t := origHookFnT.In(i)
+					v := reflect.New(t).Elem()
+					for j := 1; j < t.NumField(); j++ {
+						if j < ctxPos {
+							v.Field(j).Set(args[i].Field(j))
+						} else if j == ctxPos {
+							v.Field(j).Set(reflect.ValueOf(ctx))
+						} else {
+							v.Field(j).Set(args[i].Field(j - 1))
+						}
+					}
+					hookArgs[i] = v
+				}
+			}
+			hookResults := origHookFn.Call(hookArgs)
+			if len(hookResults) > 0 && hookResults[0].Type() == _typeOfError {
+				err, _ = hookResults[0].Interface().(error)
+			}
 
-func (la *lifecycleHookAnnotation) resolveLifecycleParamField(
-	param reflect.Value,
-	n int,
-) (
-	value reflect.Value,
-) {
-	if param.Kind() == reflect.Struct {
-		if n <= param.NumField() {
-			value = param.FieldByName(fmt.Sprintf("Field%d", n))
+			return
 		}
-	}
+		lc.Append(la.buildHook(hookFn))
+		return
+	})
 
-	return value
+	installerType := reflect.FuncOf(append(resultTypes, paramTypes...), []reflect.Type{_typeOfError}, false)
+	hookInstaller = reflect.MakeFunc(installerType, func(args []reflect.Value) (results []reflect.Value) {
+		results = []reflect.Value{_nilError}
+		var scope *dig.Scope
+		switch la.Type {
+		case _onStartHookType:
+			scope = ann.container.Scope("onStartHookScope")
+		case _onStopHookType:
+			scope = ann.container.Scope("onStopHookScope")
+		}
+		ctor := makeHookScopeCtor(paramTypes, resultTypes, args)
+		if err := scope.Provide(ctor); err != nil {
+			results[0] = reflect.ValueOf(fmt.Errorf("Error providing possible parameters for hook installer: %w", err))
+			return
+		}
+		if err := scope.Invoke(invokeFn.Interface()); err != nil {
+			results[0] = reflect.ValueOf(fmt.Errorf("Error invoking hook installer: %w", err))
+			return
```

Comment:
This return isn't technically necessary. Also, I don't think this reads as well as how you have it currently, but if you combine the ifs into an if / else if, neither return is necessary, can just rely on the one under these two ifs. Not really sure what the preference for this kind of thing is.

Reply:
You can't combine these ifs. While they look like conditionals, they're not conditionals. 

These two things are in plain English:

1. If you try providing the constructor to the scope, and it fails, set the result to an error and return
2. Then try Invoking this function, and if it fails, set the result to another error and return.

2 is dependent on 1 succeeding so it's not an if-else clause.

Reply:
I understand how conceptually you'd want to keep it how it is. But isn't this technically the same behavior:

```
if err := scope.Provide(ctor); err != nil {
    results[0] = reflect.ValueOf(fmt.Errorf("Error providing possible parameters for hook installer: %w", err))
} else if err := scope.Invoke(invokeFn.Interface()); err != nil {
    results[0] = reflect.ValueOf(fmt.Errorf("Error invoking hook installer: %w", err))
}
return
```

Currently:
1. If providing the constructor to the scope fails, we set result to an error and return.
2. Otherwise, we try to invoke the function. If this fails, we set the result to an error and return.
3. Finally, if both of those succeed, we return.

With an if / else if clause:
1. If providing the constructor to the scope fails, we set the result to an error. We do not interact with the else if, and jump straight over it to the return statement and return
2. Otherwise, we try invoking the function. If that fails, we set the result to another error. We exit the else if, and return
3. If both succeed, we return.

Aren't these the same?

Reply:
ah, I see what you mean. You're right, it is possible to remove the returns by changing this to else if. 

That being said, I don't think this necessarily helps the readability of the code. 

Here are some thoughts behind that:

1. if-else statements are usually mutually exclusive statements. i.e.
```
if A { 
  // something 
} else if B {
  // another thing
} else {
  // something else 
}
```
In the code above, we take "A/something", "B/another thing", and "C/something else" are all mutually exclusive statements that don't seem correlated. So if we grouped this together, it makes it _seem_ like Provide and Invoke are mutually exclusive.

2. That being said, you could argue that the thing that corresponds to A/B/C above in this code is _failure cases_ , which technically are mutually exclusive - you can't fail in both cases. 

3. With these in mind, the code here is doing a sequence of actions, while checking for failures in each action. It's more natural to not group them under an if-else. 

You can see examples of code like this in the standard library as well where they check for errors and return early if an error is encountered using series of if statements instead of an if-else statement. Example: https://github.com/golang/go/blob/master/src/internal/poll/fd_unix.go#L369.

Reply:
Ah okay. Yeah, I agree it's more readable how it is now. Cool, thanks for the standard library example.